src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+    }
+
+    private Object deleteKvAndGetV(RubyHash hash, int hashPos, int keyKvPos) {
+        Object deletedValue = kvStore[keyKvPos + 1];
+        index[hashPos + 1] = INDEX_SLOT_DELETED_MARKER;
+        // TODO: Instead of naively nulling out the key-value, which can produce long gaps of nulls in the kvStore,
+        //       See if we can annotate each gap with its length so that iteration code can "jump" over it
+        kvStore[keyKvPos] = null;
+        kvStore[keyKvPos + 1] = null;
+
+        hash.size--;
+        return deletedValue;
+    }
+
+    /** @return the position of the first non-null key from the end of array, or a negative value if all keys are
+     *         null */
+    private int getLastKeyPos() {
+        int lastKeyPos = kvStoreInsertionPos - 2;
+        while (lastKeyPos > 0 && kvStore[lastKeyPos] == null) {
+            lastKeyPos -= 2;
+        }
+        return lastKeyPos;
+    }
+
+    /** @param h a hash code
+     * @param max the length of the index array
+     * @return An even index ranging from 0 to max-1 */
+    private static int getIndexPosFromHash(int h, int max) {
+        return h & (max - 2);
+    }
+
+    // Or -1 if key doesn't exist
+    @GenerateUncached
+    abstract static class GetKeyPosInKVStoreNode extends RubyBaseNode {
+        public abstract int execute(Object key, int hash, int hashPosInIndex, boolean compareByIdentity,
+                int[] index, Object[] kvStore);
+
+        @Specialization
+        protected int getKeyPos(
+                Object key, int hash, int hashPosInIndex, boolean compareByIdentity, int[] index, Object[] kvStore,
+                @Cached CompareHashKeysNode compareHashKeysNode) {
+            if (hashPosInIndex == -1) {
+                return -1;
+            }
+            int kvPos = index[hashPosInIndex + 1] - 1;
+            Object otherKey = kvStore[kvPos];
+
+            // identity implies equality
+            if (otherKey == key) {
+                return kvPos;
+            }
+            // but non-identity requires we test equality
+            int otherHash = index[hashPosInIndex];
+            if (compareHashKeysNode.execute(compareByIdentity, key, hash, otherKey, otherHash)) {
+                return kvPos;
+            }
+
+            // Key doesn't exist, another one whose hash collides with its hash does
+            return -1;
+        }
+    }
+
+    // Or -1 if hash doesn't exist
+    @GenerateUncached
+    abstract static class GetHashPosInIndexNode extends RubyBaseNode {
+        public abstract int execute(int hash, int[] index);
+
+        @Specialization
+        protected int getHashPos(int hash, int[] index) {
+            int pos = getIndexPosFromHash(hash, index.length);
+
+            boolean slotUnused = index[pos + 1] == INDEX_SLOT_UNUSED_MARKER;
+            boolean slotDeleted = index[pos + 1] == INDEX_SLOT_DELETED_MARKER;
+            int firstDeletedSlot = -1;
```

Comment:
My understanding is that this value has no relation to `INDEX_SLOT_DELETED_MARKER`, but it uses the same value. That's part of why using the same `-1` literal everywhere is making this code hard to follow. Please use symbolic names where you can to make this easier to follow.